# 1.04 Activity 1

Refer to the file `files_for_activities/file1.csv` for this exercise.

1. Load data (`file1.csv`) in a new Jupyter notebook.
2. Write the code to drop the columns 'ic4' and 'ic5' from the dataframe.
3. Use the `head()` to check the new dataframe.

### Debate

1. Is it better to fill null values with the mean or with the median?
2. Can we drop all rows with null values?

1. Load data (`file1.csv`) in a new Jupyter notebook.

In [1]:
import pandas as pd
import numpy as np

In [2]:
data1=pd.read_csv("file1.csv")

In [3]:
data1.head()

CONTROLN STATE GENDER      HV1  IC1    IC4  HVP1    IC5  POBC1  POBC2  \
0     44060    FL      M   AAA896  392  520.0     7  21975      6     16   
1     96093    IL      M   537.00  365  473.0     0  19387      1     89   
2     43333    FL      F   725.00  301  436.0     3  18837     11     17   
3     21885    NC      M  AAA1095  401  413.0     7  14014      1     74   
4    190108    FL      F   995.00  252  348.0     0  17991      5      6   

     IC2  IC3    AVGGIFT  TCODE   DOB DOMAIN  TARGET_D  
0  430.0  466  28.000000      1  1901     C2     100.0  
1  415.0  410   5.666667      0     0     T2       7.0  
2  340.0  361   4.111111      0  2501     C2       5.0  
3  407.0  399  27.277778      0  2208     T2      38.0  
4  280.0  316   6.000000     28     0     C2       5.0

2. Write the code to drop the columns 'ic4' and 'ic5' from the dataframe.

In [4]:
data1.drop(columns=["IC4","IC5"],inplace=True)

In [13]:
data1.head()

CONTROLN STATE GENDER      HV1  IC1  HVP1  POBC1  POBC2    IC2  IC3  \
0     44060    FL      M   AAA896  392     7      6     16  430.0  466   
1     96093    IL      M   537.00  365     0      1     89  415.0  410   
2     43333    FL      F   725.00  301     3     11     17  340.0  361   
3     21885    NC      M  AAA1095  401     7      1     74  407.0  399   
4    190108    FL      F   995.00  252     0      5      6  280.0  316   

     AVGGIFT  TCODE   DOB DOMAIN  TARGET_D  
0  28.000000      1  1901     C2     100.0  
1   5.666667      0     0     T2       7.0  
2   4.111111      0  2501     C2       5.0  
3  27.277778      0  2208     T2      38.0  
4   6.000000     28     0     C2       5.0

#### Extra: just to check from the class

In [11]:
#Find null values method 1
data1.isnull()

CONTROLN  STATE  GENDER    HV1    IC1   HVP1  POBC1  POBC2    IC2  \
0        False  False   False  False  False  False  False  False  False   
1        False  False   False  False  False  False  False  False  False   
2        False  False   False  False  False  False  False  False  False   
3        False  False   False  False  False  False  False  False  False   
4        False  False   False  False  False  False  False  False  False   
...        ...    ...     ...    ...    ...    ...    ...    ...    ...   
1002     False  False    True  False  False  False  False  False  False   
1003     False  False   False  False  False  False  False  False  False   
1004     False  False   False  False  False  False  False  False  False   
1005     False  False   False  False  False  False  False  False  False   
1006     False  False   False  False  False  False  False  False  False   

        IC3  AVGGIFT  TCODE    DOB  DOMAIN  TARGET_D  
0     False    False  False  False   False     False  
1     False    False  False  False   False     False  
2     False    False  False  False   False     False  
3     False    False  False  False   False     False  
4     False    False  False  False   False     False  
...     ...      ...    ...    ...     ...       ...  
1002  False    False  False  False   False     False  
1003  False    False  False  False   False     False  
1004  False    False  False  False   False     False  
1005  False    False  False  False   False     False  
1006  False    False  False  False   False     False  

[1007 rows x 15 columns]

In [12]:

data1.isnull().sum()

CONTROLN     0
STATE        0
GENDER      40
HV1          2
IC1          0
HVP1         0
POBC1        0
POBC2        0
IC2          1
IC3          0
AVGGIFT      0
TCODE        0
DOB          0
DOMAIN       0
TARGET_D     0
dtype: int64

In [15]:
#Percentage of null values in the column
round(data1.isnull().sum()/len(data1),4)*100

CONTROLN    0.00
STATE       0.00
GENDER      3.97
HV1         0.20
IC1         0.00
HVP1        0.00
POBC1       0.00
POBC2       0.00
IC2         0.10
IC3         0.00
AVGGIFT     0.00
TCODE       0.00
DOB         0.00
DOMAIN      0.00
TARGET_D    0.00
dtype: float64

There is 3.97% of missing values in the column "gender" and 0.10% in the column "IC2". Although it is a small percentage we don´t know of the extent of the relevance of these values. If it were a number we could compute the missing value by subtituting the missing value with the mean, median,...

# 1.04 Activity 2

Refer to the file `files_for_activities/merged_clean_ver1.csv` for this exercise.

1. Import the data from `merged_clean_ver1.csv`. Look at the file in MS Excel to check the structure of the data. Drop the unwanted column(s).

2. Lambda Expression:

   - Create a simple lambda expression to add three numbers. Take those three numbers as input from the user. (_Since you will accept only numbers as valid inputs, check this [example](https://www.tutorialspoint.com/How-can-we-read-inputs-as-integers-in-Python) to see how to do it._)
   - Define a list as `lst = [1,2,3,4,5,6,7,8,10]`. Write a lambda expression to find the cube of a number. Use that lambda expression to find the cube of every number in the list. Define a _list comprehension_ for this question.

In [127]:
data2=pd.read_csv("merged_clean_ver1.csv")

In [128]:
data2.head()

Unnamed: 0      id state gender  median_home_val  median_household_income  \
0           0   44060    FL      M      1158.437751                      392   
1           1   96093    IL      M       537.000000                      365   
2           2   43333    FL      F       725.000000                      301   
3           3   21885    NC      M      1158.437751                      401   
4           4  190108    FL      F       995.000000                      252   

     ic2  ic3    ic4      ic5    avggift  target_d  
0  430.0  466  520.0  21975.0  28.000000     100.0  
1  415.0  410  473.0  19387.0   5.666667       7.0  
2  340.0  361  436.0  18837.0   4.111111       5.0  
3  407.0  399  413.0  14014.0  27.277778      38.0  
4  280.0  316  348.0  17991.0   6.000000       5.0

2. Lambda Expression:

   - Create a simple lambda expression to add three numbers. Take those three numbers as input from the user. (_Since you will accept only numbers as valid inputs, check this [example](https://www.tutorialspoint.com/How-can-we-read-inputs-as-integers-in-Python) to see how to do it._)
   - Define a list as `lst = [1,2,3,4,5,6,7,8,10]`. Write a lambda expression to find the cube of a number. Use that lambda expression to find the cube of every number in the list. Define a _list comprehension_ for this question.

In [34]:
inputs= lambda x,y,z:x+y+z
inputs (1,2,3)

6

In [36]:
lst =[i for i in range(1,11) ]
lst

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [50]:
lst_cube=[]
for i in lst:
    lst_cube.append(i**3)
lst_cube

[1, 8, 27, 64, 125, 216, 343, 512, 1000]

In [49]:
#list comprehension method
new_list=[i**3 for i in lst]
new_list

[1, 8, 27, 64, 125, 216, 343, 512, 1000]

# 1.04 Activity 3

Try to convert these functions into lambda functions.

```python
def function_1(a,b):
    c=a+b
    return c
```

```python
def function_2(a,b):
    c=0
    for i in range(a):
        c+=2*b
    return c
```

```python
def function_3(a,b):
    c=0
    if a>3:
        c=12
    else:
        c='Too big.'
    return c
```

In [52]:
#Function 1
def function_1(a,b):
    c=a+b
    return c
function_1(1,2)

3

In [56]:
functionl_1= lambda a,b: a+b
functionl_1(1,2)

3

In [60]:
#Function 2
def function_2(a,b):
    c=0
    for i in range(a):
        c+=2*b
    return c

function_2(1,5)

10

In [65]:
functionl_2=lambda a,b:2*a*b
functionl_2(1,5)

10

In [59]:
#Function 3
def function_3(a,b):
    c=0
    if a>3:
        c=12
    else:
        c='Too big.'
    return c

function_3(3,4)

'Too big.'

In [66]:
functionl_3= lambda a,b: 12 if a>3 else "Too big."
functionl_3(3,4)

'Too big.'

# 1.04 Activity 4

Use the same notebook in which you already loaded `merged_clean_ver1.csv`.

- Create a user-defined method to clean the column `state` in the dataframe.
- Use string functions to standardize the states to uppercase and use the strip function to clean the strings as well.



In [129]:
data2.head()

Unnamed: 0      id state gender  median_home_val  median_household_income  \
0           0   44060    FL      M      1158.437751                      392   
1           1   96093    IL      M       537.000000                      365   
2           2   43333    FL      F       725.000000                      301   
3           3   21885    NC      M      1158.437751                      401   
4           4  190108    FL      F       995.000000                      252   

     ic2  ic3    ic4      ic5    avggift  target_d  
0  430.0  466  520.0  21975.0  28.000000     100.0  
1  415.0  410  473.0  19387.0   5.666667       7.0  
2  340.0  361  436.0  18837.0   4.111111       5.0  
3  407.0  399  413.0  14014.0  27.277778      38.0  
4  280.0  316  348.0  17991.0   6.000000       5.0

In [130]:
data2["state"].unique()

array(['FL', 'IL', 'NC', 'TX', 'CA', 'NV', 'Cali', 'AP', 'MI', 'OK', 'AR',
       'IN', 'MT', 'WI', 'MO', 'HI', 'UT', 'GA', 'WA', 'ID', 'CT', 'AL',
       'ND', 'SC', 'IA', 'CO', 'LA', 'OR', 'SD', 'TN', 'NM', 'AZ', 'MN',
       'KY', 'NJ', 'NE', 'KS', 'California', 'MS', 'NY', 'Arizona', 'WY',
       'Tennessee', 'MD', 'AK', 'VA', 'AE', 'AA', 'PA', 'VT', 'WV', 'GU'],
      dtype=object)

In [133]:
data2["state"]= data2["state"].replace(["Cali","California","Arizona","Tennessee"],["CA","CA","AZ","TN"]) 
  

In [134]:
data2.state.unique()

array(['FL', 'IL', 'NC', 'TX', 'CA', 'NV', 'AP', 'MI', 'OK', 'AR', 'IN',
       'MT', 'WI', 'MO', 'HI', 'UT', 'GA', 'WA', 'ID', 'CT', 'AL', 'ND',
       'SC', 'IA', 'CO', 'LA', 'OR', 'SD', 'TN', 'NM', 'AZ', 'MN', 'KY',
       'NJ', 'NE', 'KS', 'MS', 'NY', 'WY', 'MD', 'AK', 'VA', 'AE', 'AA',
       'PA', 'VT', 'WV', 'GU'], dtype=object)

In [135]:
data2.head()

Unnamed: 0      id state gender  median_home_val  median_household_income  \
0           0   44060    FL      M      1158.437751                      392   
1           1   96093    IL      M       537.000000                      365   
2           2   43333    FL      F       725.000000                      301   
3           3   21885    NC      M      1158.437751                      401   
4           4  190108    FL      F       995.000000                      252   

     ic2  ic3    ic4      ic5    avggift  target_d  
0  430.0  466  520.0  21975.0  28.000000     100.0  
1  415.0  410  473.0  19387.0   5.666667       7.0  
2  340.0  361  436.0  18837.0   4.111111       5.0  
3  407.0  399  413.0  14014.0  27.277778      38.0  
4  280.0  316  348.0  17991.0   6.000000       5.0